plenoptic/simulate/models/naive.py

```diff
@@ -28,4 +38,180 @@ def forward(self, img):
+        f &= f/f.sum()
+
+    The signs of center and surround are determined by `center` argument. The standard
+    deviation of the surround Gaussian is constrained to be at least `width_ratio_limit`
+    times that of the center Gaussian.
+
+    Parameters
+    ----------
+    kernel_size:
+        Shape of convolutional kernel.
+    on_center:
+        Dictates whether center is on or off; surround will be the opposite of center
+        (i.e. on-off or off-on).
+    width_ratio_limit:
+        Sets a lower bound on the ratio of `surround_std` over `center_std`.
+        The surround Gaussian must be wider than the center Gaussian in order to be a
+        proper Difference of Gaussians. `surround_std` will be clamped to `ratio_limit`
+        times `center_std`.
+    amplitude_ratio:
+        Ratio of center/surround amplitude. Applied before filter normalization.
+    center_std:
+        Standard deviation of circular Gaussian for center.
+    surround_std:
+        Standard deviation of circular Gaussian for surround. Must be at least
+        `ratio_limit` times `center_std`.
+    pad_mode:
+        Padding for convolution, defaults to "circular".
+    """
+
+    def __init__(
+        self,
+        kernel_size: Union[int, Tuple[int, int]],
+        on_center: bool = True,
+        width_ratio_limit: float = 4.0,
+        amplitude_ratio: float = 1.25,
+        center_std: float = 1.0,
+        surround_std: float = 4.0,
+        pad_mode: str = "circular",
+    ):
+        super().__init__()
+
+        assert width_ratio_limit > 1.0, "stdev of surround must be greater than center"
+        assert amplitude_ratio >= 1.0, "ratio of amplitudes must at least be 1."
+
+        self.on_center = on_center
+        self.kernel_size = kernel_size
+        self.width_ratio_limit = width_ratio_limit
+        self.register_buffer("amplitude_ratio", torch.tensor(amplitude_ratio))
+
+        self.center_std = nn.Parameter(torch.tensor(center_std))
+        self.surround_std = nn.Parameter(torch.tensor(surround_std))
+
+        self.pad_mode = pad_mode
+
+    @property
+    def filt(self) -> Tensor:
+        """Creates an on center/off surround, or off center/on surround conv filter"""
+        filt_center = circular_gaussian2d(self.kernel_size, self.center_std)
+        filt_surround = circular_gaussian2d(self.kernel_size, self.surround_std)
+        on_amp = self.amplitude_ratio
+
+        if self.on_center:  # on center, off surround
+            filt = on_amp * filt_center - filt_surround  # on center, off surround
+        else:  # off center, on surround
+            filt = on_amp * filt_surround - filt_center
+
+        filt = filt / filt.sum()
+
+        return filt
+
+    def _clamp_surround_std(self) -> Tensor:
+        """Clamps surround standard deviation to ratio_limit times center_std"""
+        return self.surround_std.clamp(
+            min=self.width_ratio_limit * float(self.center_std), max=None
```

Comment:
does clamping in this way change `surround_std` in place? Or will this method return a value different than the actual `surround_std` Parameter and, if so, can that mess up learning?

Reply:
It does in-place yes, but is ok for learning. It's similar in spirit to how people do weight clipping. The value is clamped first, then the forward pass will happen, and backprop will make adjustments to the clamped value just fine.